File: gto-support-compat/src/test/java/org/ccci/gto/android/common/compat/util/LocaleCompatTest.java
package org.ccci.gto.android.common.compat.util;

import com.google.common.collect.ImmutableMap;

import org.ccci.gto.android.common.compat.util.LocaleCompat.Compat;
import org.ccci.gto.android.common.compat.util.LocaleCompat.FroyoCompat;
import org.ccci.gto.android.common.compat.util.LocaleCompat.LollipopCompat;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;
import java.util.Locale;
import java.util.Map;

import static org.junit.Assert.assertEquals;

@RunWith(Parameterized.class)
public class LocaleCompatTest {
    @Parameterized.Parameters
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][] {{new FroyoCompat()}, {new LollipopCompat()}});
    }

    private final Compat compat;

    private static final Map<Locale, String> LANGUAGETAGS;
    static {
        final ImmutableMap.Builder<Locale, String> builder = ImmutableMap.builder();
        builder.put(Locale.US, "en-US");
        builder.put(new Locale("EN", "gb"), "en-GB");
        builder.put(Locale.ENGLISH, "en");
        builder.put(Locale.SIMPLIFIED_CHINESE, "zh-CN");
        LANGUAGETAGS = builder.build();
    }

    private static final Map<String, Locale> LOCALES;
    static {
        final ImmutableMap.Builder<String, Locale> builder = ImmutableMap.builder();
        builder.put("en-US", Locale.US);
        builder.put("en-GB", Locale.UK);
        builder.put("en", Locale.ENGLISH);
        builder.put("EN-us", Locale.US);
        LOCALES = builder.build();
    }

    public LocaleCompatTest(final Compat compat) {
        this.compat = compat;
    }

    @Test
    public void testForLanguageTag() throws Exception {
        for (final Map.Entry<String, Locale> entry : LOCALES.entrySet()) {
            assertEquals(entry.getValue(), compat.forLanguageTag(entry.getKey()));
        }
    }

    @Test
    public void testToLanguageTag() throws Exception {
        for (final Map.Entry<Locale, String> entry : LANGUAGETAGS.entrySet()) {
            assertEquals(entry.getValue(), compat.toLanguageTag(entry.getKey()));
        }
    }
}

File: gto-support-compat/src/test/java/org/ccci/gto/android/common/compat/util/LocaleCompatTest.kt
package org.ccci.gto.android.common.compat.util

import androidx.test.ext.junit.runners.AndroidJUnit4
import java.util.Locale
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.hamcrest.Matchers.equalToIgnoringCase
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.Config

private val LANGUAGETAGS = mapOf<Locale, String>(
    Locale.US to "en-US",
    Locale("EN", "gb") to "en-GB",
    Locale.ENGLISH to "en",
    Locale.SIMPLIFIED_CHINESE to "zh-CN",
    Locale("es", "419") to "es-419"
)

private val LOCALES = mapOf<String, Locale>(
    "en-US" to Locale.US,
    "en-GB" to Locale.UK,
    "en" to Locale.ENGLISH,
    "EN-us" to Locale.US,
    "es-419" to Locale("es", "419")
)

@RunWith(AndroidJUnit4::class)
@Config(sdk = [21, 24])
internal class LocaleCompatTest {
    @Test
    fun testForLanguageTag() {
        LOCALES.forEach { (key, value) -> assertEquals(value, LocaleCompat.forLanguageTag(key)) }
    }

    @Test
    fun testForLanguageTagWithScript() {
        with(LocaleCompat.forLanguageTag("zh-Hans-CN")) {
            assertThat(this, equalTo(Locale.Builder().setLanguage("zh").setScript("Hans").setRegion("CN").build()))
            assertEquals("zh", language)
            assertEquals("Hans", script)
            assertEquals("CN", country)
        }
    }

    @Test
    fun testForLanguageTagWithExtension() {
        with(LocaleCompat.forLanguageTag("en-x-US")) {
            assertEquals("en", language)
            assertEquals("", country)
            assertThat(getExtension('x'), equalToIgnoringCase("US"))
        }
    }

    @Test
    fun testToLanguageTag() {
        LANGUAGETAGS.forEach { (key, value) -> assertEquals(value, LocaleCompat.toLanguageTag(key)) }
    }

    @Test
    fun testToLanguageTagWithScript() {
        assertEquals("zh-Hans-CN", LocaleCompat.toLanguageTag(Locale.forLanguageTag("zh-Hans-CN")))
    }
}
